File: ios/PteIMUIKit/Sources/PteIMUIKit/PteIMUIMediaPlayback.swift
import AVFoundation

/**
 A UIKit-wide exclusive media coordinator. Starting a voice or video playback
 stops the current one first, including playback owned by another chat screen.
 */
@MainActor
public final class PteIMUIMediaPlayback {
  public static let shared = PteIMUIMediaPlayback()
  private var activeIdentifier: String?
  private var voicePlayer: AVPlayer?
  private weak var videoPlayer: AVPlayer?

  private init() {}

  public func playVoice(identifier: String, url: URL) {
    if activeIdentifier == identifier, voicePlayer != nil { stop(); return }
    stop()
    let player = AVPlayer(url: url)
    activeIdentifier = identifier
    voicePlayer = player
    player.play()
  }

  public func activateVideo(identifier: String, player: AVPlayer) {
    if activeIdentifier != identifier { stop() }
    activeIdentifier = identifier
    videoPlayer = player
  }

  public func release(identifier: String) {
    guard activeIdentifier == identifier else { return }
    clear()
  }

  public func stop() {
    voicePlayer?.pause()
    videoPlayer?.pause()
    clear()
  }

  private func clear() {
    voicePlayer = nil
    videoPlayer = nil
    activeIdentifier = nil
  }
}

File: android/pte-im-uikit/src/main/java/com/ptelive/im/ui/PteIMUIMediaPlayback.kt
package com.ptelive.im.ui

import android.content.Context
import android.media.AudioAttributes
import android.media.MediaPlayer

/**
 * UIKit-wide exclusive media playback coordinator.  Message cells and preview
 * pages route through it so a newly started voice/video always stops the old
 * one, even when they were created by different conversations.
 */
object PteIMUIMediaPlayback {
  private var activeKey: String? = null
  private var voicePlayer: MediaPlayer? = null
  private var stopVideo: (() -> Unit)? = null

  @Synchronized
  fun playVoice(context: Context, key: String, source: String) {
    if (source.isBlank()) return
    if (activeKey == key && voicePlayer != null) { stop(); return }
    stop()
    val player = MediaPlayer()
    runCatching {
      player.setAudioAttributes(AudioAttributes.Builder().setContentType(AudioAttributes.CONTENT_TYPE_SPEECH).setUsage(AudioAttributes.USAGE_MEDIA).build())
      player.setDataSource(source)
      player.setOnPreparedListener { it.start() }
      player.setOnCompletionListener { release(key) }
      player.setOnErrorListener { _, _, _ -> release(key); true }
      activeKey = key
      voicePlayer = player
      player.prepareAsync()
    }.onFailure { player.release(); clear() }
  }

  @Synchronized
  fun activateVideo(key: String, stop: () -> Unit) {
    if (activeKey != key) stop()
    activeKey = key
    stopVideo = stop
  }

  @Synchronized
  fun release(key: String) {
    if (activeKey != key) return
    voicePlayer?.release()
    clear()
  }

  @Synchronized
  fun stop() {
    voicePlayer?.run { stop(); release() }
    stopVideo?.invoke()
    clear()
  }

  private fun clear() { voicePlayer = null; stopVideo = null; activeKey = null }
}
